### PR TITLE
Project automation CRUD

### PR DESCRIPTION
**Pull request by dave-shawley**
I wanted to throw this up for review since it is a departure from the existing codebase in a number of ways. This is creating a new entity in the system called an _Automation_ that is associated with an existing _Integration_ and available for use in one or more _Project Types_. So... the "gitlab" Integration would have an Automation named "create-project" that is available for whichever project types have a gitlab project associated to them.  The relationship between entities is something like:
```mermaid
erDiagram
    automations }o--|| integrations : "owned by"
    automations }o--|{ project_types : "available for"
    automations }o--o{ automations : "requires"
```
The "available for" and "requires" relationships are separate tables. The link between the Automation and its "owning" Integration is a simple foreign key.

Now for the departure part ... instead of creating a hierarchical API structure like:

| Endpoint | Description |
| --- | --- |
| POST /integrations/{name}/automations | create a new automation |
| POST /integrations/{name}/automations/{name}/project-types | make it available to a specific project type |
| POST /integrations/{name}/automations/{name}/dependencies | make it require another automation be enabled |

with the accompanying `DELETE` and `PUT` methods for full CRUD. I opted to go a different route and use more complex documents. So creating a new automation is a `POST` with something that looks like:
```json
{
  "name": "create-project",
  "applies_to": ["api", "consumer"],
  "depends_on": []
}
```
This requires insertion into a handful of tables along with some magic to make it work with either IDs or slugs in those arrays. The response is a handful of `LEFT JOIN`s as well.

So that is one change. The other is the shift to pydantic for validation (currently in addition to OpenAPI). There are a handful of reasons for this but the conciseness of the application code is what I really like. It is nice that pydantic implements `ImportString` as a type that handles dynamically importing a callable 🤯  What I really liked was how it cleaned up the Postgres gymnastics required for arrays of enumerated values and what not. It also allowed me to jettison the SQL blobs and related class attributes (eg, `NAME`, `ITEM_NAME`, `ID_KEY`, etc). The resulting collection `GET` method ended up being:
```python
class CollectionRequestHandler(base.PydanticHandlerMixin,
                               base.ValidatingRequestHandler):
    async def get(self, integration_name: str) -> None:
        result = await self.postgres_execute(
            'SELECT a.id, a.name, a.slug, a.callable, a.categories,'
            '       a.integration_name, array_agg(pt.slug) AS applies_to,'
            '       array_agg(DISTINCT d.slug) AS depends_on'
            '  FROM v1.automations AS a'
            '  LEFT JOIN v1.available_automations AS aa'
            '         ON aa.automation_id = a.id'
            '  LEFT JOIN v1.project_types AS pt ON pt.id = aa.project_type_id'
            '  LEFT JOIN v1.automations_graph AS g ON g.automation_id = a.id'
            '  LEFT JOIN v1.automations AS d ON d.id = g.dependency'
            ' WHERE a.integration_name = %(integration_name)s'
            ' GROUP BY a.id, a.name, a.slug, a.callable, a.categories,'
            '          a.integration_name, a.slug',
            parameters={'integration_name': integration_name})
        self.send_response([Automation.model_validate(row) for row in result])
```
with `POST` not being much more difficult.

Anyway, there is a lot to digest in here but I would like some feedback on the approach.
* _Is it too magical?_
* _Is it too different?_
* If we keeping the pydantic stuff, should they move into `imbi.models` or keep the models with the handlers?
* Does inlining the SQL bother anyone greatly?

### CODE REVIEW

File: imbi/slugify.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def path_element(value: str) -> str:
+    """Generate a slug for an arbitrary string"""
+    value = re.sub(r'[\s_]+', '-', value)
+    value = re.sub(r'[^-a-z0-9]', '', value.lower())
+    return Slug(value.strip('-_'))
+
+
+def decode_path_slug(slug_or_id: str) -> tuple[int | None, str | None]:
```
Comment (in-op) [inline imbi/slugify.py:22]: I think the output typing would actually be

```python
tuple[int, None] | tuple[None, str]
```

**Comment by in-op** [comment]
**Is it too magical?**
I don't think so - everything is still explicit, like defining the models

**Is it too different?**
I think there are benefits to the validation, and it'd be really nice if we could generate openapi docs from it... although I know that can be complicated and not worth...

I do think the `NAME` field on the endpoints still needs to be there tho, we increment a stat using it.

**If we keeping the pydantic stuff, should they move into imbi.models or keep the models with the handlers?**
I think `models` should probably be a directory, not just a file, and we keep the main models in there. The only ones I see making more sense in the endpoints are the request models, they feel like they make more sense there.

**Does inlining the SQL bother anyone greatly?**
I don't mind it inline, but the triple quotes are nice because PyCharm will syntax highlight the SQL, but doesn't with a series of individual strings (it really _should_ though).

**Comment by dave-shawley** [comment]
> **Does inlining the SQL bother anyone greatly?** I don't mind it inline, but the triple quotes are nice because PyCharm will syntax highlight the SQL, but doesn't with a series of individual strings (it really _should_ though).

Interesting ... I wonder if there is a toggle for that somewhere since it works for me 🤔 

image

**Comment by in-op** [comment]
Ah, weird, it doesn't recognize it when I make a SQL string anew. I have to re-index the project...

File: imbi/slugify.py (hunk 1 above)
Comment (dave-shawley) [reply imbi/slugify.py:35]: doh ... missed this one

File: imbi/endpoints/integrations/automations.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        The `slug` can be a slug, the name of the automation,
+        or the automation's surrogate ID.
+
+        """
+        automation_id, automation_slug = slugify.decode_path_slug(slug)
+        del slug  # use either automation_id or automation_slug
+
+        self.logger.debug('looking for %r or %r', automation_id,
+                          automation_slug)
+        result = await self.postgres_execute(
```
[the remark this answers is not in the document]
Comment (dave-shawley) [reply imbi/endpoints/integrations/automations.py:331]: Added missing cases in [cb6e9d6](https://github.com/AWeber-Imbi/imbi-api/pull/65/commits/cb6e9d6d91784915a00c67ca55e87f8fab398a38)

**Pull request merged**